backend/src/main/java/com/google/sps/FindMatchQuery.java

```diff
@@ -114,29 +107,65 @@ private MatchPreference getCombinedMatchPreference(
    */
   private boolean isCombinedMatchPreferenceSatisfied(
       MatchPreference combinedMatchPreference,
-      Participant newParticipant,
-      Participant currParticipant) {
+      Participant firstParticipant,
+      Participant secondParticipant) {
     if (combinedMatchPreference == MatchPreference.ANY) {
+      // both ANY, doesn't matter how many same fields
       return true;
     }
 
-    // Count number of same fields between the two participants
-    boolean sameRole = newParticipant.getRole().equals(currParticipant.getRole());
-    boolean sameProductArea =
-        newParticipant.getProductArea().equals(currParticipant.getProductArea());
-    int numSameFields = Booleans.countTrue(sameRole, sameProductArea);
-    System.out.println("number of same fields: " + numSameFields);
+    // Get lists of combined role, product area, interests for first and second participant
+    String firstFilledFieldsList = getAllFilledFieldsList(firstParticipant);
+    String secondFilledFieldsList = getAllFilledFieldsList(secondParticipant);
+    System.out.println("First filled fields: " + firstFilledFieldsList);
+    System.out.println("Second filled fields: " + secondFilledFieldsList);
+
+    // Count number of shared fields
+    StringTokenizer firstFields = new StringTokenizer(firstFilledFieldsList, ",");
+    StringTokenizer secondFields = new StringTokenizer(secondFilledFieldsList, ",");
+    int firstNumFilledFields = firstFields.countTokens();
+    int secondNumFilledFields = secondFields.countTokens();
+    int numSameFields = 0;
+    while (firstFields.hasMoreTokens()) {
+      String firstField = firstFields.nextToken();
+      secondFields = new StringTokenizer(secondFilledFieldsList, ",");
+      while (secondFields.hasMoreTokens()) {
+        if (firstField.equals(secondFields.nextToken())) {
+          numSameFields++;
+          System.out.println(firstField + " matches");
+        }
+      }
+    }
+    System.out.println("numSameFields: " + numSameFields);
 
-    // If match preference = SIMILAR, numSameFields must be at least MIN_SAME_FIELDS to satisfy
-    // preference
+    // Check if match based on preference and number of same fields
+    int maxNumFilledFields = Math.max(firstNumFilledFields, secondNumFilledFields);
+    int MIN_SAME_FIELDS = maxNumFilledFields / 2 + maxNumFilledFields % 2;
+    System.out.println(MIN_SAME_FIELDS);
     if (combinedMatchPreference == MatchPreference.SIMILAR && numSameFields < MIN_SAME_FIELDS) {
       return false;
     }
-    // If match preference = DIFFERENT, numSameFields must be less than MIN_SAME_FIELDS to satisfy
-    // preference
     if (combinedMatchPreference == MatchPreference.DIFFERENT && numSameFields >= MIN_SAME_FIELDS) {
       return false;
     }
+    System.out.println("combined match pref satisfied");
     return true;
   }
+
+  /** Return all filled fields of participant in one string delimited by , */
```

Comment:
It's worth a mention here that this function assumes none of the Role, PA or interests should have the same words in other categories. Otherwise, this match might fail